File: app/src/main/java/com/microjet/airqi2/GestureLock/DefaultPatternSettingActivity.java
package com.microjet.airqi2.GestureLock;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.widget.TextView;

import com.github.ihsg.patternlocker.OnPatternChangeListener;
import com.github.ihsg.patternlocker.PatternIndicatorView;
import com.github.ihsg.patternlocker.PatternLockerView;
import com.microjet.airqi2.Definition.SavePreferences;
import com.microjet.airqi2.R;

import java.util.List;

public class DefaultPatternSettingActivity extends AppCompatActivity {

    private PatternIndicatorView patternIndicatorView;
    private TextView textMsg;
    private PatternHelper patternHelper;

    public static void startAction(Context context) {
        Intent intent = new Intent(context, DefaultPatternSettingActivity.class);
        context.startActivity(intent);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_default_pattern_setting);

        this.patternIndicatorView = findViewById(R.id.pattern_indicator_view);
        PatternLockerView patternLockerView = findViewById(R.id.pattern_lock_view);
        this.textMsg = findViewById(R.id.text_msg);

        patternLockerView.setOnPatternChangedListener(new OnPatternChangeListener() {
            @Override
            public void onStart(PatternLockerView view) {
            }

            @Override
            public void onChange(PatternLockerView view, List<Integer> hitList) {
            }

            @Override
            public void onComplete(PatternLockerView view, List<Integer> hitList) {
                boolean isOk = isPatternOk(hitList);
                view.updateStatus(!isOk);
                patternIndicatorView.updateState(hitList, !isOk);
                updateMsg();
            }

            @Override
            public void onClear(PatternLockerView view) {
                finishIfNeeded();
            }
        });

        this.textMsg.setText(getResources().getText(R.string.text_set_pattern));
        this.patternHelper = new PatternHelper();
    }

    private boolean isPatternOk(List<Integer> hitList) {
        this.patternHelper.validateForSetting(hitList);
        return this.patternHelper.isOk();
    }

    private void updateMsg() {
        this.textMsg.setText(this.patternHelper.getMessage());
        this.textMsg.setTextColor(this.patternHelper.isOk() ?
                getResources().getColor(R.color.colorPrimary) :
                getResources().getColor(R.color.colorAccent));
    }

    private void finishIfNeeded() {
        if (this.patternHelper.isFinish()) {
            finish();

            SharedPreferences share = getSharedPreferences(SavePreferences.SETTING_KEY, Context.MODE_PRIVATE);
            share.edit().putBoolean(SavePreferences.SETTING_MAP_PRIVACY, true).apply();
        }
    }
}

File: app/src/main/java/com/microjet/airqi2/GestureLock/DefaultPatternSettingActivity.kt
package com.microjet.airqi2.GestureLock

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.support.v7.app.AppCompatActivity
import com.github.ihsg.patternlocker.OnPatternChangeListener
import com.github.ihsg.patternlocker.PatternIndicatorView
import com.github.ihsg.patternlocker.PatternLockerView
import com.microjet.airqi2.Definition.SavePreferences
import com.microjet.airqi2.R
import kotlinx.android.synthetic.main.activity_default_pattern_setting.*

class DefaultPatternSettingActivity : AppCompatActivity() {

    private var patternHelper: PatternHelper? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_default_pattern_setting)

        patternLockerView.setOnPatternChangedListener(object : OnPatternChangeListener {
            override fun onStart(view: PatternLockerView) {}

            override fun onChange(view: PatternLockerView, hitList: List<Int>) {}

            override fun onComplete(view: PatternLockerView, hitList: List<Int>) {
                val isOk = isPatternOk(hitList)
                view.updateStatus(!isOk)
                //patternIndicatorView!!.updateState(hitList, !isOk)
                updateMsg()
            }

            override fun onClear(view: PatternLockerView) {
                finishIfNeeded()
            }
        })

        text_msg.text = resources.getText(R.string.text_set_pattern)
        this.patternHelper = PatternHelper()
    }

    private fun isPatternOk(hitList: List<Int>): Boolean {
        this.patternHelper!!.validateForSetting(hitList)
        return this.patternHelper!!.isOk
    }

    private fun updateMsg() {
        text_msg.text = this.patternHelper!!.message
        text_msg.setTextColor(if (this.patternHelper!!.isOk)
            ContextCompat.getColor(this, R.color.Main_textResult_Good)
        else
            ContextCompat.getColor(this, R.color.Main_textResult_Bad))
    }

    private fun finishIfNeeded() {
        if (this.patternHelper!!.isFinish) {
            finish()

            val share = getSharedPreferences(SavePreferences.SETTING_KEY, Context.MODE_PRIVATE)
            share.edit().putBoolean(SavePreferences.SETTING_MAP_PRIVACY, true).apply()
        }
    }

    companion object {

        fun startAction(context: Context) {
            val intent = Intent(context, DefaultPatternSettingActivity::class.java)
            context.startActivity(intent)
        }
    }
}